From ff92553e817e6afee089123c3e772132a3f54b11 Mon Sep 17 00:00:00 2001
From: Marcin Niestroj <m.niestroj@emb.dev>
Date: Tue, 29 Jul 2025 15:54:49 +0200
Subject: [PATCH] gateway: support device cert upload

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>
---
 include/golioth/gateway.h | 11 +++++++++++
 src/gateway.c             | 21 +++++++++++++++++++++
 2 files changed, 32 insertions(+)

diff --git a/include/golioth/gateway.h b/include/golioth/gateway.h
index cf233373..893ce9d2 100644
--- a/include/golioth/gateway.h
+++ b/include/golioth/gateway.h
@@ -78,6 +78,17 @@ enum golioth_status golioth_gateway_server_cert_get(struct golioth_client *clien
                                                     size_t *len,
                                                     int32_t timeout_s);
 
+/// Set device certificate.
+///
+/// @param client The client handle from @ref golioth_client_create
+/// @param buf Pointer to buffer where certificate is stored
+/// @param len Certificate length
+/// @param timeout_s Timeout in seconds for API call
+enum golioth_status golioth_gateway_device_cert_set(struct golioth_client *client,
+                                                    const void *buf,
+                                                    size_t len,
+                                                    int32_t timeout_s);
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/src/gateway.c b/src/gateway.c
index 2d2f6433..7ebb9553 100644
--- a/src/gateway.c
+++ b/src/gateway.c
@@ -199,4 +199,25 @@ enum golioth_status golioth_gateway_server_cert_get(struct golioth_client *clien
     return GOLIOTH_OK;
 }
 
+enum golioth_status golioth_gateway_device_cert_set(struct golioth_client *client,
+                                                    const void *buf,
+                                                    size_t len,
+                                                    int32_t timeout_s)
+{
+    uint8_t token[GOLIOTH_COAP_TOKEN_LEN];
+    golioth_coap_next_token(token);
+
+    return golioth_coap_client_set(client,
+                                   token,
+                                   ".g/device-cert",
+                                   "",
+                                   GOLIOTH_CONTENT_TYPE_OCTET_STREAM,
+                                   buf,
+                                   len,
+                                   NULL,
+                                   NULL,
+                                   true,
+                                   timeout_s);
+}
+
 #endif  // CONFIG_GOLIOTH_GATEWAY
-- 
2.50.1

